core/include/moveit/task_constructor/stages/deep_grasp_pose.h

```diff
@@ -0,0 +1,310 @@
+	void setGraspPose(const std::string& grasp) { properties().set("grasp", grasp); }
+	void setGraspPose(const moveit_msgs::RobotState& grasp) { properties().set("grasp", grasp); }
+
+protected:
+	void onNewSolution(const SolutionBase& s) override;
+
+private:
+	bool found_candidates_;
+	std::vector<geometry_msgs::PoseStamped> grasp_candidates_;
+	std::vector<double> costs_;
+};
+
+template <class ActionSpec>
+DeepGraspPose<ActionSpec>::DeepGraspPose(const std::string& action_name, const std::string& stage_name,
+                                         double goal_timeout, double server_timeout)
+  : GeneratePose(stage_name), ActionBaseT(action_name, false, goal_timeout, server_timeout), found_candidates_(false) {
+	auto& p = properties();
+	p.declare<std::string>("eef", "name of end-effector");
+	p.declare<std::string>("object");
+	p.declare<double>("angle_delta", 0.1, "angular steps (rad)");
+
+	p.declare<boost::any>("pregrasp", "pregrasp posture");
+	p.declare<boost::any>("grasp", "grasp posture");
+
+	ROS_INFO_NAMED(LOGNAME, "Waiting for connection to grasp generation action server...");
+	ActionBaseT::clientPtr_->waitForServer(ros::Duration(ActionBaseT::server_timeout_));
+	ROS_INFO_NAMED(LOGNAME, "Connected to server");
+}
+
+template <class ActionSpec>
+void DeepGraspPose<ActionSpec>::composeGoal() {
+	Goal goal;
+	goal.action_name = ActionBaseT::action_name_;
+	ActionBaseT::clientPtr_->sendGoal(
+	    goal, std::bind(&DeepGraspPose<ActionSpec>::doneCallback, this, std::placeholders::_1, std::placeholders::_2),
+	    std::bind(&DeepGraspPose<ActionSpec>::activeCallback, this),
+	    std::bind(&DeepGraspPose<ActionSpec>::feedbackCallback, this, std::placeholders::_1));
+
+	ROS_INFO_NAMED(LOGNAME, "Goal sent to server: %s", ActionBaseT::action_name_.c_str());
+}
+
+template <class ActionSpec>
+bool DeepGraspPose<ActionSpec>::monitorGoal() {
+	// monitor timeout
+	bool monitor_timeout = ActionBaseT::goal_timeout_ > std::numeric_limits<double>::epsilon() ? true : false;
+	double timeout_time = ros::Time::now().toSec() + ActionBaseT::goal_timeout_;
+
+	while (ActionBaseT::nh_.ok()) {
+		ros::spinOnce();
+
+		if (found_candidates_) {
+			break;
+		}
+
+		if (ros::Time::now().toSec() > timeout_time && monitor_timeout) {
+			ActionBaseT::clientPtr_->cancelGoal();
+			ROS_ERROR_NAMED(LOGNAME, "Grasp pose generator time out reached");
+			return false;
+		}
+	}
+
+	return true;
+}
+
+template <class ActionSpec>
+void DeepGraspPose<ActionSpec>::activeCallback() {
+	ROS_INFO_NAMED(LOGNAME, "Generate grasp goal now active");
+	found_candidates_ = false;
+}
+
+template <class ActionSpec>
+void DeepGraspPose<ActionSpec>::feedbackCallback(const FeedbackConstPtr& feedback) {
+	// each candidate should have a cost
+	if (feedback->grasp_candidates.size() != feedback->costs.size()) {
+		ROS_ERROR_NAMED(LOGNAME, "Invalid input: each grasp candidate needs an associated cost");
+	} else {
```

Comment:
Rather than the extra `else` block, you could just return here:
```suggestion
          return;
	}
```
(Then also remove the `}`.)

Reply:
The reason I don't return here is I thought the error and info messages provide the user with feedback to let them know if the action message fields where completed properly.

Reply:
I don't mean you should remove the error, but you can return immediately after it, instead of using an `else` block. It would be functionally equivalent to what you have implemented.
```
template <class ActionSpec>
void DeepGraspPose<ActionSpec>::feedbackCallback(const FeedbackConstPtr& feedback) {
	// each candidate should have a cost

	if (feedback->grasp_candidates.size() != feedback->costs.size()) {
		ROS_ERROR_NAMED(LOGNAME, "Invalid input: each grasp candidate needs an associated cost");
		return;
	}
	ROS_INFO_NAMED(LOGNAME, "Grasp generated feedback received %lu candidates: ", feedback->grasp_candidates.size());

	grasp_candidates_.resize(feedback->grasp_candidates.size());
	costs_.resize(feedback->costs.size());

	grasp_candidates_ = feedback->grasp_candidates;
	costs_ = feedback->costs;

	found_candidates_ = true;
}
```

Reply:
I see that does simplify it. I will make that change.